### PR TITLE
IH-577 Implement v7 UUID generation

### PR DESCRIPTION
**Pull request by rdn32**
* New function Uuidx.make_v7_uuid, with the idea being that ordering v7 UUIDs alphabetically will also order them by creation time
* The values produced by Uuidx.make_uuid_urnd hadn't necessarily been valid UUIDs, since the variant and version fields were being filled in randomly - this is now fixed so that it returns v4 UUIDs.
* There are a couple of functions for generating v4 and v7 from known inputs, for the purpose of unit testing. (The v4 function is mainly there so I could check the setting of variant and version fields by comparing the output with that which Python's UUID module produces.)

### CODE REVIEW

File: ocaml/libs/uuid/uuidx.mli [hunk 1]
```diff
@@ -27,14 +27,26 @@
...
 
 val make_uuid_urnd : unit -> 'a t
 
+val make_v4_uuid : int64 -> int64 -> 'a t
+(** For testing only: Create a v4 UUID, as defined in RFC 9562 5.4 *)
+
+val make_v7_uuid_from_time_and_bytes : int * int64 -> int64 -> 'a t
+(** For testing only: create a v7 UUID, as defined in RFC 9562 5.7 *)
+
+val make_v7_uuid : unit -> 'a t
```
Comment (psafont) [inline ocaml/libs/uuid/uuidx.mli:43]: Could this be changed to accept a function for the timestamp? That way we could use a monotonic clock source. The signature could be `(unit -> uint64) -> 'a t`
I know it would not be standard-compliant, and we would need to be careful about not persisting the produced timestamps, but I believe losing monotonicity here is something very undesired.

Comment (rdn32) [reply ocaml/libs/uuid/uuidx.mli:43]: I've gone for a hybrid approach whereby I use ptime to get the system time initially, then use mtime to get monotonic adjustments to that - so that we keep monotonicity but still have timestamps that are at least approximately correct.

**Comment by lindig** [comment]
Is there a chance to upstream this code? We are using `uuidm` and it would be nice to expose this to more scrutiny by upstreaming it.

**Comment by psafont** [comment]
> Is there a chance to upstream this code? We are using uuidm and it would be nice to expose this to more scrutiny by upstreaming it.

Because it's a new standard, it has leeway on how it can be implemented, and it needs to be as an additional package, I think there's a low chance to get into upstream. Having the suggestion upstream would maybe make its author to implement it

**Comment by rdn32** [comment]
It looks like generating a timestamp can't reasonably be done by uuidm (without introducing a dependency) but constructing v7 UUID given a timestamp could be. I can put together a PR. and see how well it goes down.

**Comment by rdn32** [comment]
> It looks like generating a timestamp can't reasonably be done by uuidm (without introducing a dependency) but constructing v7 UUID given a timestamp could be. I can put together a PR. and see how well it goes down.

The PR is https://github.com/dbuenzli/uuidm/pull/14. [Daniel Bünzli](https://github.com/dbuenzli) has added a '👍' reaction, which I take to be a good sign.

**Comment by psafont** [comment]
He usually commits his own version of the change while referencing the original author

**Comment by dbuenzli** [comment]
Please chime in [on the PR](https://github.com/dbuenzli/uuidm/pull/14) to discuss this. I think it's a good idea to add basic support for v7 in uuidm.

**Comment by lindig** [comment]
I understand that v7 UUID will be implemented upstream as I suggested - which is great. Could you summarize what this PR achieves?

**Comment by rdn32** [comment]
> I understand that v7 UUID will be implemented upstream as I suggested - which is great. Could you summarize what this PR achieves?

The main thing it does is to give uuidx a wrapper for uuidm's new v7 UUID generation - including producing monotonic-yet-broadly-accurate timestamps.

**Review by lindig — APPROVED** [review]

File: ocaml/libs/uuid/uuidx.ml [hunk 2]
```diff
@@ -131,21 +131,43 @@ let read_bytes dev n =
...
+  let start = Mtime_clock.counter () in
+  let t0 =
+    let d, ps = Ptime_clock.now () |> Ptime.to_span |> Ptime.Span.to_d_ps in
+    Int64.(add (mul (of_int d) 86_400_000_000_000L) (div ps 1000L))
+  in
+  fun () ->
+    let since_t0 = Mtime_clock.count start |> Mtime.Span.to_uint64_ns in
+    Int64.add t0 since_t0
+
+let make_v7_uuid () = make_v7_uuid_from_parts (now_ns ()) (rand64 ())
```
Comment (psafont) [inline ocaml/libs/uuid/uuidx.ml:169]: There are only 2048 possible valued for the rand64, how are we avoiding duplicates?

In general, I'd prefer to use a counter here, it would ensure that no duplicates are ever produced, and that sequentially-created uuids are properly ordered.

However, the rates needed for this to happen are rather high, and I doubt this affects how the toolstack will use the uuids, this can be changed at a later time if we found it's an impediment.

**Review by psafont — APPROVED** [review]

**Comment by lindig** [comment]
Can we run a `make format` on this to satisfy the checks?

File: ocaml/libs/uuid/dune [hunk 3]
```diff
@@ -3,8 +3,13 @@
   (public_name uuid)
   (modules uuidx)
   (libraries
-    unix (re_export uuidm)
+    mtime
+    mtime.clock.os
```
Comment (dbuenzli) [inline ocaml/libs/uuid/dune:7]: Note that both `{ptime,mtime}.clock.os` are deprecated use `{ptime,mtime}.os` if you are using the latest version (also clocks are made to re-export the base package so you can also specify only those if you wish so). See https://discuss.ocaml.org/t/ann-ptime-1-2-0-mtime-2-1-0-qrc-0-2-0/15268

File: ocaml/libs/uuid/uuidx.ml (hunk 2 above)
Comment (rdn32) [reply ocaml/libs/uuid/uuidx.ml:169]: > There are only 2048 possible valued for the rand64, how are we avoiding duplicates?

I'd been assuming that `Random.State.bits64` would give 2^64 different values. We're only using 62 of its bits, but that still gives us about 4 * 10^18 different values, which I would have thought we could rely on chance to avoid getting duplicate values within 245 nanoseconds of each other. Or is the `Random` module very much worse than I'd imagined?
